### ISSUE TITLE
1.Data gathering: crawl data from Reddit.

### ISSUE THREAD

**Issue by tmozgach** [tmozgach/ent_ob#1]
*

**Comment by neowangkkk** [comment]
Hi,

Glad to form a work team here. The first job is that we want to crawl more data from the target community. 
https://www.reddit.com/r/Entrepreneur/

The time period we want to crawl is from 2012.1.1 to 2017.12.31. 

The required variables are shown in the attachment. If you have any questions, please let me know.
![screen shot 2018-01-09 at 1 22 38 pm](https://user-images.githubusercontent.com/6750790/34743950-745f745c-f540-11e7-9688-39247c2acefc.png)

[Sample.xlsx](https://github.com/tmozgach/ent_ob/files/1616745/Sample.xlsx)

Best,
Tao

**Comment by tmozgach** [comment]
@neowangkkk  I haven't done it before, so in order to save time and be consistence could you please give me the script or method/web link how you did it before. Is that like that? https://www.labnol.org/internet/web-scraping-reddit/28369/

**Comment by neowangkkk** [comment]
@tmozgach Last time I paid $150 to hire a part-time programmer to crawl the data. He told me he used C+ language to write the program. I don't believe he will give me the code:-( 

I am not sure the difficulty level of doing web crawling at reddit.com. Can you please search and check if any python package or sth else can do it? If after 20 work hours it is still a problem, we may go to the part-time programmer again. I understand some websites put a lot of tricks to prevent people crawling their content. It may be a huge task that only people with years of experience in crawling can handle. But it is worthy for you to learn and try when our time still allows.

The google sheet method in your link may have some flaws. It said sub-reddit can only show 1000 posts. But our last crawling got over 25,000 threads for 18 months.

If you have any questions, please feel free to let me know.

**Comment by tmozgach** [comment]
**1) Option: PRAW API** 

0) Install `pip` and `draw` without root:
https://gist.github.com/saurabhshri/46e4069164b87a708b39d947e4527298
```
curl -L https://bootstrap.pypa.io/get-pip.py -o getpip.py
python get-pip.py --user
python -m pip install --user praw
```
1)  Reddit video tutorials:
https://www.youtube.com/watch?v=NRgfgtzIhBQ

Note:
`user_agent` parameter? 
```
Archival Bot
```
```
According to Reddit the user agent is required to determine basic information about the script accessing Reddit, such as the name, the version and the author (and that's supposed to go in there, such as "prawtutorial v1.0 by /u/sentdex" or similar).
```

```

r = praw.Reddit(client_id = '***',
                     client_secret = '***',
                     user_agent = 'Archival Bot')```


Things to try:
https://www.reddit.com/r/botwatch/comments/3i0s4c/how_to_get_all_posts_from_a_subreddit/
https://github.com/peoplma/subredditarchive/blob/master/subredditarchive.py
https://www.reddit.com/r/help/comments/799cfp/how_can_i_find_all_posts_in_a_subreddit_between
Research more on '~ 1.7 billion comments dataset'

**Comment by tmozgach** [comment]
https://pushshift.io/using-bigquery-with-reddit-data/

**Comment by neowangkkk** [comment]
Wow. That's great! Very promising. Can we output the query result into R or python data format?

**Comment by tmozgach** [comment]
neowangkkk probably there is a way, but I faced with another the issue, in BigQuery they store comments and POSTS for 2016,2017 year, but there is NO posts for 2015,2014,2013,2012, only comments! I mean, there is no main posts, title (from sender), only their comments (replies). 
1) I am wondering which year(data) do you have already?
2) There is no tables with information about the author, so we can not count the karma correctly because information about posts for 2012-2015 years don't exist . Are you sure that by now we need this information about author in order to conduct Topic Modeling?
3) I asked one data science about data, he said that  gathering data correctly and properly and cleaning it will required 1 month full time job for such big period in the best case. So, could we decrease period? Why 2012? May be we can just gather posts and comments for 2016,2017 and only comments for 2012,2013,2014,2015 and try on them Topic Modeling? Without Karma and etc by now?
4) If we really really need information about comments for 2012-2015 year, I can try to crawl it by script. But it seems that only gathering information and concatenation will take A LOT of time.

**Comment by tmozgach** [comment]
@neowangkkk I will try to crawl old data and see if we can reach data by script. I will text you later about result and we can decide what should we do.

**Comment by tmozgach** [comment]
* Parse JSON to CSV.
1) Output the JSON file nicely:
```
python -m json.tool my_json.json
```

**Comment by tmozgach** [comment]
Raw information for 2012 - 201 years:
https://drive.google.com/open?id=1S8iAnnyjH4NxZpXmkX0ppN9c_Hm6XJLQ

**Comment by tmozgach** [comment]
@neowangkkk 
TPostComRaw.csv: Not clean titles, main posts and comments for 2012-2017 years.
TPostRaw.csv: Not clean titles and main posts WITHOUT comments for 2012-2017 years.
https://drive.google.com/drive/folders/1S8iAnnyjH4NxZpXmkX0ppN9c_Hm6XJLQ?usp=sharing

**Comment by tmozgach** [comment]
Probably I need to join comments, post and title as ONE paragraph?

**Comment by neowangkkk** [comment]
Yes. As discussed last time, combining all texts in one thread may generate better outcome in clustering/topic modelling. Please go ahead and try it.

In addition, I checked the old data of karma. It is a fixed value for each individual at the point of crawling. We can’t get the changing karma through the time. So later if you can get the karma data for the participants in your investigated period, that would be fine.

**Comment by tmozgach** [comment]
Every row of that csv  is one thread (title, post, comments).

**Comment by tmozgach** [comment]
[2009_2011data.csv.zip](https://github.com/tmozgach/ent_ob/files/1798209/2009_2011data.csv.zip)

**Comment by tmozgach** [comment]
2009 - 2017 all data:

https://www.dropbox.com/s/tlq6gfnnlnqvumx/data0312.csv?dl=0

**Comment by tmozgach** [comment]
Data for Topic Modeling 2009-2017:
https://drive.google.com/drive/folders/1S8iAnnyjH4NxZpXmkX0ppN9c_Hm6XJLQ?usp=sharing
newRawAllData.csv

**Comment by tmozgach** [comment]
```
library(tidyverse)
myData = read_delim("/home/tatyana/nlp/data0312.csv", delim = '\t' )
library(zoo)

library(dplyr)
newD = data.frame(col1 = myData$Replier, col2 = myData$`Points from this question`) %>% 
  
  do(na.locf(.))

write_csv(newD, "dataForPyth.csv")




```

**Comment by tmozgach** [comment]
Latest data:
https://www.dropbox.com/s/50vkf5makcojd5w/data_full.csv?dl=0
